### ISSUE TITLE
cargo raze should update the Cargo.lock file

### ISSUE THREAD

**Issue by djmarcin** [google/cargo-raze#309]
Currently, I have to manually run `cargo generate-lockfile` in order to add sha256s after adding new dependencies to Cargo.toml. It seems like `cargo raze` should be doing this automatically.

**Comment by UebelAndre** [comment]
I agree that raze should be better about this. There's https://github.com/google/cargo-raze/pull/222 already which is an attempt to solve this. I have a branch with some WIP stuff I'd like to open a PR for if nothing else has been done by the time I get around to it (waiting for other PRs to be merged first).

Some things to consider:
- I don't think raze should be constantly generating new lock files every run. It should attempt to use what's there or be able to identify when the lockfile is outdated or incomplete. 
- Because of the custom logic for handling binary dependencies (see https://github.com/google/cargo-raze/issues/218) which outputs a `Cargo.raze.lock` file, the `Cargo.lock` file can't always be the source of truth. So I'm not sure how best to ensure the correct lockfile is up to date in that context.

My first attempt at this was to try and make the `Cargo.raze.lock` file always generated and always the authority and ran into some inconsistencies between remote and vendored genmodes. Again, if I can get some other PRs closed out then I'll be able to fix up my branch and open a PR.

**Comment by dae** [comment]
For now I've found it helpful to invoke cargo-raze via a wrapper: https://github.com/ankitects/anki/blob/master/cargo/update.py

**Comment by UebelAndre** [comment]
> For now I've found it helpful to invoke cargo-raze via a wrapper: https://github.com/ankitects/anki/blob/master/cargo/update.py

Question, why are you using a custom requests BUILD file? Is there something that can't be expressed via the raze settings?

**Comment by dae** [comment]
To use separate features on Linux: #234

**Comment by UebelAndre** [comment]
Cargo raze now supports a `--generate-lockfile` argument but there is one caveat in that it will generate a `Cargo.raze.lock` file. If one is present in your workspace it will take precedence over the traditional `Cargo.lock` file. Would this help here?

**Comment by djmarcin** [comment]
> Cargo raze now supports a `--generate-lockfile` argument but there is one caveat in that it will generate a `Cargo.raze.lock` file. If one is present in your workspace it will take precedence over the traditional `Cargo.lock` file. Would this help here?

One annoyance we had with `cargo generate-lockfile` is that it seems to ignore any existing lockfile and just update everything to the latest (semver compatible) version. This is bad because any time anyone adds a new dependency they have to deal with dozens of simultaneous dependency version changes unrelated to their new dependency.

Does raze's `--generate-lockfile` option have the same behavior?

**Comment by UebelAndre** [comment]
It does. `--generate-lockfile` for cargo raze is effectively the equivalent command for cargo. And yeah, it'll output a `Cargo.raze.lock` file currently. This is something I hate and have a great desire to change. It's important for all users of `binary_deps` but I think with either some smarter alternatives or work to Cargo, we can satisfy the same need and not have to maintain something unique outside of Cargo. You point out something I've never really thought of. If I add a dependency, how do I add it to my lock file without updating all my dependencies? Is there really no way to do this in Cargo? Is there an RFC for it?

**Comment by briansmith** [comment]
This probably should be a new issue, but one thing that's important for my use is for cargo-raze to have a `--locked` mode where it will fail with an error if there Cargo.lock doesn't exist with all the correct info. My use case is that I'm migrating a large project to Bazel and the current build system is in production use and already ensures that the Cargo.lock file has exactly the right contents. I need to regularly re-run cargo-raze to sync the Bazel build system with the production one while we compare them head-to-head.

> Currently, I have to manually run cargo generate-lockfile in order to add sha256s after adding new dependencies to Cargo.toml. It seems like cargo raze should be doing this automatically.

In a large project with lots of dependencies one almost never can/should run `cargo generate-lockfile` and expect reasonable results. In projects I've worked on we have a rule that we only update specific dependencies using `cargo update -p <dependency-name>` and we're never supposed to run `cargo update` without `-p`, which seems similar to `cargo generate-lockfile`. This is partly done to reduce merge conflicts, and partly because builds becomes unreliable when dependencies are accidentally updated through `cargo update`/`cargo generate-lockfile`.

I suggest that if it is important to have raze try to handle *new* dependencies (ones without digests) then the scope of this request should be limited to just that, and otherwise the user can/should use `cargo` to make other kinds of changes. I also suggest that there be a separate invocation, e.g. `cargo raze --update-new-dependencies` for doing this, and that `cargo raze` never updates the Cargo.lock unless explicitly requested to do so.
